input-explainer.md

```diff
@@ -177,22 +177,34 @@ Most applications will want to visually represent the input sources somehow. The
+
+Controller devices should generally only be rendered if the `XRSession`'s `environmentBlendMode` is `'opaque'`, as any other mode implies that the user can see any physical device they may be holding. Rendering cursors or target rays may still be desirable.
+
+If the input source is a handheld device, the `XRInputSource` object should have a non-null `gamepad` attribute. The `Gamepad`'s `id` is used to determine what should be rendered if the app intends to visualize the input source itself, rather than an alternative virtual object. (See the section on [Button and Axis State](#button-and-axis-state) for more details.)
+
+The WebXR Device API currently does not offer any way to retrieve renderable resources that represent the input devices from the API itself, and as such the `Gamepad`'s `id` must be used as a key to load an appropriate resources from the application's server or a CDN. The example below presumes that the `getControllerMesh` call would do the required lookup and caching.
+
 ```js
 // These methods presumes the use of a fictionalized rendering library.
 
 // Render a visualization of the input source - eg. a controller mesh.
 function renderInputSource(xrFrame, inputSource) {
-  // Don't render an input source if it doesn't have a grip space.
-  if (!inputSource.gripSpace)
+  // Don't render an input source if it doesn't have a grip space, doesn't have
+  // a gamepad, or if the blend mode is not 'opaque'. 
+  if (!inputSource.gripSpace || !inputSource.gamepad ||
+      xrFrame.session.environmentBlendMode != 'opaque')
     return;
 
-  let gripPose = xrFrame.getPose(inputSource.gripSpace, xrFrameOfRef);
+  // Retrieve a mesh to render based on the gamepad object's id. If no mesh is
+  // found skip rendering.
+  let controllerMesh = getControllerMesh(inputSource.gamepad.id);
+  if (!controllerMesh)
+    return;
 
-  // Controller meshes should not be rendered on transparent displays (AR), so
-  // only render a controller mesh if the XREnvironmentBlendMode is 'opaque' and
-  // we have a valid gripPose to transform it with.
-  if (gripPose && xrFrame.session.environmentBlendMode == 'opaque') {
-    let controllerMesh = getControllerMesh(inputSource);
+  // Only render a controller mesh if we have a valid gripPose.
+  let gripPose = xrFrame.getPose(inputSource.gripSpace, xrFrameOfRef);
+  if (gripPose) {
     renderer.drawMeshAtTransform(controllerMesh, gripPose.transform);
```

Comment:
When you rebase this, can we switch from straight-up rendering the mesh to just saving it off for the imaginary renderer to do?  I think there may be an old commit hanging around somewhere that I'd done this if you want me to dig it up...

Reply:
Hey! This just so happens to be what I did anyway! Yay!